content/advent-2017/messaging-framework.md

```diff
@@ -0,0 +1,709 @@
+	consumer.Consume()
+
+	signal_chan := make(chan os.Signal, 1)
+	signal.Notify(signal_chan,
+		syscall.SIGINT,
+		syscall.SIGTERM,
+		syscall.SIGQUIT,
+		syscall.SIGKILL)
+
+	go func() {
+		for {
+			s := <-signal_chan
+			switch s {
+			case syscall.SIGINT:
+				fmt.Println("syscall.SIGINT")
+				cleanup()
+				// kill -SIGTERM XXXX
+			case syscall.SIGTERM:
+				fmt.Println("syscall.SIGTERM")
+				cleanup()
+				// kill -SIGQUIT XXXX
+			case syscall.SIGQUIT:
+				fmt.Println("syscall.SIGQUIT")
+				cleanup()
+			case syscall.SIGKILL:
+				fmt.Println("syscall.SIGKILL")
+				cleanup()
+			default:
+				fmt.Println("Unknown signal.")
+			}
+		}
+	}()
+
+	go func(){
+		http.HandleFunc("/", handler)
+		http.ListenAndServe(":8080", nil)
+	}()
+
+	tcpServer.Listen()
+
+
+}
+
+func handler(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprint(w, "TCP Server is up and running!")
+}
+
+func cleanup(){
+	tcpServer.Close()
+	producer.Close()
+	consumer.Close()
+	os.Exit(0)
+}
+
+
+
+func onNewConnection(clientUid string) {
+	log.Println("onNewConnection, uid: ", clientUid)
+}
+
+func onConnectionTerminated(clientUid string) {
+	log.Println("onConnectionTerminated, uid: ", clientUid)
+}
+
+/**
+Called when data is received from a TCP client, will generate a message to the message broker
+ */
+func onDataReceived(clientUid string, data []byte) {
+	log.Println("onDataReceived, uid: ", clientUid, ", data: ", string(data))
+	if string(data)=="Ping" {
+		log.Println("sending Pong")
+		//answer with pong
+		tcpServer.SendDataByClientId(clientUid, []byte("Pong"))
+	}
+	if producer!=nil {
+		var deviceRequest models.DeviceRequest
+		err:= json.Unmarshal(data,&deviceRequest)
+		if err==nil {
+			serverRequest := models.ServerRequest{
+				DeviceRequest: deviceRequest,
+				ServerId: "1",
+				ClientId: clientUid,
+			}
+			producer.Produce(serverRequest)
+		} else {
+			log.Println(err)
+		}
+
+	}
+
+}
+
+func onProducerError(err error){
+	log.Println("onProducerError: ", err)
+}
+
+func onConsumerError(err error){
+	log.Println("onConsumerError: ",err)
+}
+
+func onDataConsumed(data []byte){
+	log.Println("onDataConsumed: ", string(data))
+	var serverResponse models.ServerResponse
+	err := json.Unmarshal(data,&serverResponse)
+	if err!=nil {
+		log.Println(err)
+		return
+	}
+	if serverResponse.DeviceResponse.Action == "connect.response" && serverResponse.DeviceResponse.Status == "ok" && serverResponse.ClientId!= "" {
+		//attach the device id to our existing client
+		err =tcpServer.SetDeviceUidToClient(serverResponse.ClientId,serverResponse.DeviceUid)
+		if err!=nil {
+			log.Println(err)
+		}
+	}
+	toSend, err := json.Marshal(serverResponse.DeviceResponse)
+	if err!=nil {
+		log.Println(err)
+		return
+	}
+	if serverResponse.ClientId!="" {
+		tcpServer.SendDataByClientId(serverResponse.ClientId,toSend)
+	} else {
+		if serverResponse.DeviceUid!=""{
+			tcpServer.SendDataByDeviceUid(serverResponse.DeviceUid,toSend)
+		}
+	}
+
+
+}
+```
```

Comment:
This is too much code - link to the source and explain important parts, but leave the whole listing out, people will not read it.

Reply:
do you refer to a specific section or it's a general comment?